### PR TITLE
Address edge cases and add tests

### PR DESCRIPTION
**Pull request by dan4thewin**
This PR solves 3 edge cases and adds unit tests to prevent future regressions.

### CODE REVIEW

File: source/core_json.c [hunk 1]
```diff
@@ -51,6 +51,12 @@ typedef union
     ( ( ( x ) == ' ' ) || ( ( x ) == '\t' ) || \
       ( ( x ) == '\n' ) || ( ( x ) == '\r' ) )
 
+#define isOpenBracket_( x )     ( ( ( x ) == '{' ) || ( ( x ) == '[' ) )
+#define isCloseBracket_( x )    ( ( ( x ) == '}' ) || ( ( x ) == ']' ) )
+/* NB. The numeric values of the open and close bracket pairs differ by 2. */
```
Comment (aggarw13) [inline source/core_json.c:56]: What does "NB" stand for?

File: test/unit-test/core_json_utest.c [hunk 2]
```diff
@@ -717,10 +738,6 @@ void test_JSON_Validate_Partial_Documents( void )
     jsonStatus = JSON_Validate( CUT_AFTER_OBJECT_START_MARKER,
                                 CUT_AFTER_OBJECT_START_MARKER_LENGTH );
     TEST_ASSERT_EQUAL( JSONPartial, jsonStatus );
-
-    jsonStatus = JSON_Validate( CUT_AFTER_KEY,
```
Comment (aggarw13) [inline test/unit-test/core_json_utest.c:721]: Why was this removed?

File: source/core_json.c (hunk 1 above)
Comment (dan4thewin) [reply source/core_json.c:56]: Nota bene

File: test/unit-test/core_json_utest.c (hunk 2 above)
Comment (dan4thewin) [reply test/unit-test/core_json_utest.c:721]: With the changes to the parser, a key without a value is considered illegal, not partial.  The case is also covered by one of the new tests.

**Review by yourslab — APPROVED** [review]

File: test/unit-test/core_json_utest.c (hunk 2 above)
Comment (dan4thewin) [reply test/unit-test/core_json_utest.c:721]: Also, I put the test back with the other illegal tests.

**Review by cobusve — APPROVED** [review]

File: source/core_json.c (hunk 1 above)
Comment (aggarw13) [reply source/core_json.c:56]: May be _esoteric_ for readers?

Comment (dan4thewin) [reply source/core_json.c:56]: We use e.g. and i.e. often enough.  I'll keep it.
